src/v/io/persistence.cc

```diff
@@ -0,0 +1,327 @@
+ * request to avoid this case when using Seastar's allocator in release mode.
+ *
+ * This only matters for the in-memory file system because Seastar's file system
+ * interface already has stricter requirements that avoid this case implicitly.
+ * We want to be able to use the in-memory file system to get weird with
+ * alignments and test edge cases.
+ */
+seastar::temporary_buffer<char>
+memory_persistence::allocate(uint64_t alignment, size_t size) noexcept {
+    auto ret = seastar::temporary_buffer<char>::aligned(
+      alignment, seastar::align_up(size, alignment));
+    ret.trim(size);
+    return ret;
+}
+
+seastar::future<seastar::shared_ptr<persistence::file>>
+memory_persistence::create(std::filesystem::path path) noexcept {
+    auto ret = files_.try_emplace(
+      path, seastar::make_shared<memory_file>(this));
+    if (!ret.second) {
+        return seastar::make_exception_future<
+          seastar::shared_ptr<persistence::file>>(
+          std::filesystem::filesystem_error(
+            "File exists",
+            path,
+            std::error_code(EEXIST, std::system_category())));
+    }
+    return seastar::make_ready_future<seastar::shared_ptr<persistence::file>>(
+      ret.first->second);
+}
+
+seastar::future<seastar::shared_ptr<persistence::file>>
+memory_persistence::open(std::filesystem::path path) noexcept {
+    auto it = files_.find(path);
+    if (it == files_.end()) {
+        return seastar::make_exception_future<
+          seastar::shared_ptr<persistence::file>>(
+          std::filesystem::filesystem_error(
+            "File does not exist",
+            path,
+            std::error_code(ENOENT, std::system_category())));
+    }
+    return seastar::make_ready_future<seastar::shared_ptr<persistence::file>>(
+      it->second);
+}
+
+memory_persistence::memory_file::memory_file(memory_persistence* persistence)
+  : persistence_(persistence) {}
+
+seastar::future<size_t> memory_persistence::memory_file::dma_read(
+  uint64_t pos, char* buf, size_t len) noexcept {
+    if (auto err = check_alignment(
+          "dma_read",
+          pos,
+          buf,
+          len,
+          memory_dma_alignment(),
+          disk_read_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return read(pos, buf, len);
+}
+
+seastar::future<size_t> memory_persistence::memory_file::dma_write(
+  uint64_t pos, const char* buf, size_t len) noexcept {
+    if (auto err = check_alignment(
+          "dma_write",
+          pos,
+          buf,
+          len,
+          memory_dma_alignment(),
+          disk_write_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return write(pos, buf, len).then([this, pos](size_t written) {
+        size_ = std::max(size_, pos + written);
+        return written;
+    });
+}
+
+seastar::future<> memory_persistence::memory_file::close() noexcept {
```

Comment:
Should we add assertions around nothing being used after `close` and nothing outstanding during `close`?

Reply:
ah yeh. i've added this to the todo list. it's a bit of a bigger change, but definitely should be done at some point.